### ISSUE TITLE
Is there a way to "expand" a row?

### ISSUE THREAD

**Issue by zoff** [akveo/ng2-smart-table#346]
I'm looking to ng2-smart-table to migrate my angular1 components based on ng-table (https://github.com/esvit/ng-table) to Angular (>=2).

One of the thing I use ng-table for the most is like this:
```
<table ....>
  <tr ng-repeat-start="elem in $ctrl.elements">
    <td data-title="'Actions'"><button ng-click="$ctrl.toggleDetails(elem)">Details</button></td>
    <td data-title="'Field 1'">{{elem.field1}}</td>
    <td data-title="'Field 2'">{{elem.field2}}</td>
    <td data-title="'Field 3'">{{elem.field3}}</td>
    <td data-title="'Field 4'">{{elem.field4}}</td>
  </tr>
  <tr ng-repeat-end ng-if="elem.showDetails">
    <td colspan="5">DETAILS for elem</td>
  </tr>
</table>
```

Is there a way to achieve something similar with ng2-smart-table?

**Comment by hzitoun** [comment]
@zoff have you managed to do that?

**Comment by skyazid** [comment]
@zoff I'm looking for the same thing. Have you managed to do that, please?

**Comment by zoff** [comment]
Nope

**Comment by tenebroso** [comment]
I was able to do this just a couple of days ago:  http://recordit.co/AtO9Bsz2Ns

1) Set the details data as a secondary object within your table row. 

```
{
    name: 'Janet Dawson',  
    date: 'Fri May 12 2017 11:05:20 GMT-0500 (CDT)',
    detailInfo: {
        custom1: 'ID: 348976556',
        custom2: 'Sponsor: Lily Robbins',
        }
 }
```

2) Get the DOM element of a particular row on by passing an event: (mouseover)="getRow($event)". 

```
getRow(ev) {
    // getParentsUntil is a utility function that queries the DOM for the parent element 
    // based on the selector
    this.selectedRow = this.getParentsUntil(ev.target, '.ng2-smart-row')[0];
 }
```

3) Dynamically create the new <tr> and loop through each item in the secondary object to create the new row:

```
const newRow  = document.createElement('tr');
// Loop through keys and create 'td' elements for each
// passing in the new value as the innerHTML
```

4) Append the new <tr> after the selected row:

```this.selectedRow.insertAdjacentHTML('afterend', newRow.innerHTML);```

This is how I did it, in a nutshell. Let me know if you need any more specifics.

**Comment by nicomana** [comment]
@tenebroso Hi Mate, could you plz add like a working example?. Sry I couldn't make it work.
Thx a lot!

**Comment by aefox** [comment]
I will create a PR this week to support this functionality. Hopefully the repo owners will be responsive for the review

**Comment by crissty** [comment]
@aefox I can not see you PR, did you open it?
This property would give it a lot of value to the component

**Comment by KambagiriswamyBo-Kore** [comment]
@tenebroso  Is there any workaround for this?

**Comment by Et3rnal** [comment]
I spend lots of time trying to achieve this and almost give up (I should've spent that time changing the table itself instead). But I've also learnt a lot of things.

thanks for @tenebroso for the inspiration on how this can be done. however, his method didn't work for me as I needed to load a component there and not only show data.

Please note, what I'm sharing still needs some work and I hope someone here can help me finish it. One of the things needs to be looked at is the use of [closest()](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest) which is not supported by all browsers and it doesn't return value from the first click!?

You will find a few links to pages where I found what I used.
https://stackblitz.com/edit/dynamically-row-components-for-smart-table

Please share your forks if you do any improvement.
Another simpler thing isn't there is hiding the component, there is a commented code in inji.service.ts which will destroy it after 3 seconds. but we need something like a toggle so to hide or destroy so if the user clicks multiple times doesn't create the component again.

**Comment by Et3rnal** [comment]
In case anyone interested, I have finalised the solution above, now works perfectly, the button works as a toggle and you can close it within as well.

**Comment by Alontal** [comment]
> In case anyone interested, I have finalised the solution above, now works perfectly, the button works as a toggle and you can close it within as well.

@Et3rnal look like code has a wierd bug(I tried yesterday and it was working) did you make any changes? - error: reached to maximum call stack. (see screenshot).
![image](https://user-images.githubusercontent.com/25777598/58168850-e8412480-7c97-11e9-98be-e23ead793485.png)

**Comment by Et3rnal** [comment]
@Alontal Yes, I noticed if you have <td> tag inside the inserted component it will break. It's fixed now and we can have <td> without issue.

I'm constantly improving the code as I face issues. I'm working now to make closing from within the add component notify the main adding button. it should be easy but I'm lost again :D

**Comment by Alontal** [comment]
@Et3rnal Thanks! I suggest that this should be added as a feature to the repository.

**Comment by Et3rnal** [comment]
@Alontal well looking at the number of pull requests I have a feeling the developers of this library has abandon it! 
Also, the way I implement it is without a need to change the library itself. If we want to change the code It would be a lot easier I think

**Comment by MingzhouLin** [comment]
@Et3rnal Hi, thank you for your work. But I find a issue that if you add a row there and do filter on one column on the same time, the appended row would be moved to bottom. Is there a way to bind the appended row to the row we click the button. Thank you very much.

**Comment by Et3rnal** [comment]
@MingzhouLin So many issues this way so I stopped using it. I hade the same issue when paging and I think I did listen to clicks on paging to hide any open rows. Good luck

**Comment by me-hungry** [comment]
@Et3rnal Sorting doesn't work for this [https://stackblitz.com/edit/dynamically-row-components-for-smart-table](url)

**Comment by Et3rnal** [comment]
@me-hungry Sorry mate, I didn't continue with my attempt. Good luck

**Comment by me-hungry** [comment]
> @me-hungry Sorry mate, I didn't continue with my attempt. Good luck

@Et3rnal No problem, I added remove component in OnDestroy. Also need to remove empty row(parent of component). Anyway thanks for your method. Coz no other options :(

**Comment by Eshaan7** [comment]
Was anyone able to integrate this with ng2-smart-table ?
